x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportGetShutdownStatusAction.java

```diff
@@ -111,6 +150,95 @@ protected void masterOperation(
         listener.onResponse(response);
     }
 
+    // pkg-private for testing
+    static ShutdownShardMigrationStatus shardMigrationStatus(
+        ClusterState currentState,
+        String nodeId,
+        SingleNodeShutdownMetadata.Type shutdownType,
+        AllocationDeciders allocationDeciders,
+        ClusterInfoService clusterInfoService,
+        SnapshotsInfoService snapshotsInfoService,
+        AllocationService allocationService
+    ) {
+        // Only REMOVE-type shutdowns will try to move shards, so RESTART-type shutdowns should immediately complete
+        if (SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownType)) {
+            return new ShutdownShardMigrationStatus(
+                SingleNodeShutdownMetadata.Status.COMPLETE,
+                0,
+                "no shard relocation is necessary for a node restart"
+            );
+        }
+
+        // First, check if there are any shards currently on this node, and if there are any relocating shards
+        int startedShards = currentState.getRoutingNodes().node(nodeId).numberOfShardsWithState(ShardRoutingState.STARTED);
+        int relocatingShards = currentState.getRoutingNodes().node(nodeId).numberOfShardsWithState(ShardRoutingState.RELOCATING);
+        int initializingShards = currentState.getRoutingNodes().node(nodeId).numberOfShardsWithState(ShardRoutingState.INITIALIZING);
+        int totalRemainingShards = relocatingShards + startedShards + initializingShards;
+
+        // If there's relocating shards, or no shards on this node, we'll just use the number of shards left to move
+        if (relocatingShards > 0 || totalRemainingShards == 0) {
+            SingleNodeShutdownMetadata.Status shardStatus = totalRemainingShards == 0
+                ? SingleNodeShutdownMetadata.Status.COMPLETE
+                : SingleNodeShutdownMetadata.Status.IN_PROGRESS;
+            return new ShutdownShardMigrationStatus(shardStatus, totalRemainingShards);
+        } else if (initializingShards > 0 && relocatingShards == 0 && startedShards == 0) {
+            // If there's only initializing shards left, return now with a note that only initializing shards are left
+            return new ShutdownShardMigrationStatus(
+                SingleNodeShutdownMetadata.Status.IN_PROGRESS,
+                totalRemainingShards,
+                "all remaining shards are currently INITIALIZING and must finish before they can be moved off this node"
+            );
+        }
+
+        // If there's no relocating shards and shards still on this node, we need to figure out why
+        final RoutingAllocation allocation = new RoutingAllocation(
+            allocationDeciders,
+            currentState.getRoutingNodes(),
+            currentState,
+            clusterInfoService.getClusterInfo(),
+            snapshotsInfoService.snapshotShardSizes(),
+            System.nanoTime()
+        );
+        allocation.debugDecision(true);
```

Comment:
I *think* (at least, I can't see it anywhere) that we don't use the actual message for the decision anywhere. Since we don't, I think we can remove this line and avoid the debugging decision (which allows allocation deciders to short-circuit and be a bit more efficient).

Either that, or maybe we can log the the actual decision (which will include a plethora of details) at `TRACE` level if an unmoveable shard is found? That would be nice for debugging I think?

Reply:
I'm not exactly sure why, but calling `allocationService.explainShardAllocation()` with a `RoutingAllocation` that doesn't have `debugDecision` set to `true` trips [this assertion](https://github.com/elastic/elasticsearch/blob/a79d1a5205d4bda57694d24040bb8f1b07cd68ee/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L522). I think that might just be because the only other place that calls this method is the Allocation Explain API, so we might be able to change it but I didn't want to do so without consulting the distributed team. Which could be a follow-up PR?

In the mean time, I'll change this to at least log the decision at `TRACE`.

Reply:
also, you can use:

```java
allocation.setDebugMode(DebugMode.EXCLUDE_YES_DECISIONS);
```

To only return the "NO"s, which will cut down on the amount of logging also (while still being in debug mode)